### PR TITLE
Fix source pruning when there is a keyword in resource path

### LINKED ISSUE #36128: Source pruning when there is a keyword in resource path
**Description:**

Consider the below sample

```bal
service / on new http:Listener(8080) {
    resource function get limit/[int id] () {
        
    }
}
```

Parser generated source file
```bal
service / on new http:Listener(8080) {
    resource function get  MISSING[]/[int id] () {
        
    }
}
```
Node: the `limit` keyword is removed from the source file instead of reporting it as invalid token

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #36128

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -6915,7 +6915,8 @@ private STNode parseResourcePathSegment(boolean isFirstSegment) {
         STToken nextToken = peek();
         switch (nextToken.kind) {
             case IDENTIFIER_TOKEN:
-                if (isFirstSegment && nextToken.isMissing() && getNextNextToken().kind == SyntaxKind.SLASH_TOKEN) {
+                if (isFirstSegment && nextToken.isMissing() && peek(0).kind == SyntaxKind.IDENTIFIER_TOKEN &&
+                        getNextNextToken().kind == SyntaxKind.SLASH_TOKEN) {
                     // special case `[MISSING]/` to improve the error message for `/hello`
                     consume(); // to ignore current missing identifier diagnostic
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:6921]: The fix `peek(0).kind == SyntaxKind.IDENTIFIER_TOKEN` is correct. But, we rely on the previously consumed token which is vulnerable to error if the grammar before `resource-path-segment` changes. 

The core problem here is using `consume();` to ignore the missing identifier. When we call `consume()` if there are any invalid tokens in the invalid token stack they are also attached to the leading minutiae of the `consume()` token.

Therefore, I suggest following to ignore the missing token(virtually inserted token). 
```suggestion
                    this.insertedToken = null; // to ignore current missing identifier diagnostic
```

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Now we don't need `peek(0).kind == SyntaxKind.IDENTIFIER_TOKEN` check right?. Also it's more clean to introduce `removeInsertedToken()` method in `AbstractParser` and do this.insertedToken = null there WDYT?

Comment (lasinicl) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: If we remove `peek(0).kind == SyntaxKind.IDENTIFIER_TOKEN` check, `resource path cannot begin with slash` error is also logged with the `invalid token 'limit'` error. Wouldn't that be confusing for the user?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: +1 let's keep the check

**Review by rdulmina — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Our assumption here is when the `peek(0).kind == SyntaxKind.IDENTIFIER_TOKEN`, the invalid node stack is empty.
Therefore, I prefer `this.invalidNodeInfoStack.isEmpty()` check instead.  `invalidNodeInfoStack` is private atm.  We can have something like `isInvalidNodeStackEmpty()`. WDYT?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: +1

**Pull request merged**